### PR TITLE
[PAL/Linux-SGX] Add missing EACCEPT after restricting page perms

### PR DESCRIPTION
**Pull request by boryspoplawski**
## Description of the changes 
Bugfix

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)

### CODE REVIEW

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: 0 of 2 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NM9kwHUE4gBo-WIh-T6:-NM9kwHUE4gBo-WIh-T7:b-ouvqdn) (no related file):*
This PR was not tested at all. Since I have no time for that, somebody should do it
cc @dimakuv

___

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski and @dimakuv)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NM9kwHUE4gBo-WIh-T6:-NMDcUJwBqGtVULZCa6a:bxdh81z) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

This PR was not tested at all. Since I have no time for that, somebody should do it
cc @dimakuv
</blockquote></details>

On it.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMDc_Xf0J8nn2lWAT_1:-NMDc_Xf0J8nn2lWAT_2:bw233nt) (no related file):*
Context for this bugfix: Restricting enclave page permissions must be followed by an `EACCEPT` on this page.

An excerpt from Intel SDM, Section 35.5.10 "Restricting the EPCM Permissions of a Page":

> The typical process for restricting the permissions of an enclave page is as follows:
> 1. Enclave requests that the OS to restrict the permissions of an EPC page.
> 2. OS performs permission restriction, flushing cached linear-address to physical-address translations, and pagetable modifications.
> ....
> 3. Enclave invokes the EACCEPT leaf function.
>     a. Enclave may access the page throughout the entire process.
>     b. Successful call to EACCEPT guarantees that no stale cached linear-address to physical-address translations
are present.

---

Interestingly, Borys introduced a test together with the EDMM PR, which actually tests restricting enclave page permissions: https://github.com/gramineproject/gramine/blob/ff9ec79ab27fb4f95daa35656d1410c695264509/pal/regression/memory.c#L110-L121

In this test, a page with RW permissions is restricted to R permissions, and then a subsequent write access is supposed to segfault.

I verified that our Jenkins CI correctly passes this test. I'm currently debugging this test, to see if works somehow incorrectly.

The interesting part about this flow is that Intel SDM doesn't really say that the page access will #GP... The only real consequence of EACCEPT seems to be what's described in step 3.b. The way I interpret that sentence from Intel SDM -- if there was e.g. another enclave thread that accessed the same page *before changing permissions*, it would still have the old permissions in its local CPU cache (in the TLB I guess?), so this other enclave thread will **not** notice the new permissions.

**TODO**: blocking myself. I'd like to verify this with other SGX experts.

___
___
*[`pal/src/host/linux-sgx/sgx_arch.h` line 292 at r1](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMDi6CJ6b3JQgGG09Fv:-NMDi6CKCPKgvug0Mzn2:b-ze8ndv) ([raw file](https://github.com/gramineproject/gramine/blob/176839c9b69e1b7f41fcbae6835ebe6226dc0c6f/pal/src/host/linux-sgx/sgx_arch.h#L292)):*
FYI: Can be found in Intel SDM, Vol. 3, Chapter 34.12.1 (Table 34-18).

In particular, this new `_PR` flag is:

> If 1 indicates that a permission restriction operation on the page is in progress, otherwise a permission restriction operation is not in progress.

**Comment by dimakuv** [comment]
Jenkins, retest Jenkins-Debug-18.04 please (`fdatasync01` LTP test timed out, known and unrelated)

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required) (waiting on @boryspoplawski)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NM9kwHUE4gBo-WIh-T6:-NMEEfXW8IE7Wfi5f37S:b-896fix) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

On it.
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMDc_Xf0J8nn2lWAT_1:-NMEEh_I9-kvf5hXNQLq:bp6zpp5) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Context for this bugfix: Restricting enclave page permissions must be followed by an `EACCEPT` on this page.

An excerpt from Intel SDM, Section 35.5.10 "Restricting the EPCM Permissions of a Page":

> The typical process for restricting the permissions of an enclave page is as follows:
> 1. Enclave requests that the OS to restrict the permissions of an EPC page.
> 2. OS performs permission restriction, flushing cached linear-address to physical-address translations, and pagetable modifications.
> ....
> 3. Enclave invokes the EACCEPT leaf function.
>     a. Enclave may access the page throughout the entire process.
>     b. Successful call to EACCEPT guarantees that no stale cached linear-address to physical-address translations
are present.

---

Interestingly, Borys introduced a test together with the EDMM PR, which actually tests restricting enclave page permissions: https://github.com/gramineproject/gramine/blob/ff9ec79ab27fb4f95daa35656d1410c695264509/pal/regression/memory.c#L110-L121

In this test, a page with RW permissions is restricted to R permissions, and then a subsequent write access is supposed to segfault.

I verified that our Jenkins CI correctly passes this test. I'm currently debugging this test, to see if works somehow incorrectly.

The interesting part about this flow is that Intel SDM doesn't really say that the page access will #GP... The only real consequence of EACCEPT seems to be what's described in step 3.b. The way I interpret that sentence from Intel SDM -- if there was e.g. another enclave thread that accessed the same page *before changing permissions*, it would still have the old permissions in its local CPU cache (in the TLB I guess?), so this other enclave thread will **not** notice the new permissions.

**TODO**: blocking myself. I'd like to verify this with other SGX experts.
</blockquote></details>

Unblocking myself. I verified with one SGX hardware expert, and my understanding (explained above) is correct.

Basically, this bug could have been found under a very specific set of conditions: several enclave threads + data race on accessing the restricted-permissions page. This is very hard to reproduce, so I wouldn't bother with writing tests. Approving.

___

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMDc_Xf0J8nn2lWAT_1:-NMKJ8GT5wnzIFlQPaAl:b-zb8qnv) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Unblocking myself. I verified with one SGX hardware expert, and my understanding (explained above) is correct.

Basically, this bug could have been found under a very specific set of conditions: several enclave threads + data race on accessing the restricted-permissions page. This is very hard to reproduce, so I wouldn't bother with writing tests. Approving.
</blockquote></details>

I don't understand your explanation. What it has to do with this test? Why didn't it fail? Shouldn't the page have still the old permissions until EACCEPTed?

___

**Review by dimakuv — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required) (waiting on @boryspoplawski)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMDc_Xf0J8nn2lWAT_1:-NMSzSaU59Ynna2zoNd2:beff06a) (no related file):*
> Shouldn't the page have still the old permissions until EACCEPTed?

No, the page has the new permissions on this logical processor (the processor which called an OCALL to ask the host OS to perform `EMODPR`). But the local TLBs of other logical processors may still contain the old permissions. And the EACCEPT verifies/forces a TLB flush on this enclave page (some microarchitectural details).

___

**Review by mkow — DISMISSED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: all files reviewed, 3 unresolved discussions (waiting on @boryspoplawski)

___
*[`pal/src/host/linux-sgx/enclave_edmm.c` line 75 at r1](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMT2L47-0-cBL13L6mg:-NMT2L47-0-cBL13L6mh:undefined) ([raw file](https://github.com/gramineproject/gramine/blob/176839c9b69e1b7f41fcbae6835ebe6226dc0c6f/pal/src/host/linux-sgx/enclave_edmm.c#L75)):*
> ```C
>                               | SGX_SECINFO_FLAGS_PR | prot);
>             if (ret < 0) {
>                 log_error("%s: failed to accept restrict pages permissions at %#lx: %d", __func__,
> ```





_Suggestion:_
```C
restricted pages
```
___
*[`pal/src/host/linux-sgx/enclave_edmm.c` line 140 at r1](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMT2MgnAfzdVLJQEoFT:-NMT2MgnAfzdVLJQEoFU:undefined) ([raw file](https://github.com/gramineproject/gramine/blob/176839c9b69e1b7f41fcbae6835ebe6226dc0c6f/pal/src/host/linux-sgx/enclave_edmm.c#L140)):*
> ```C
>                                                 | SGX_SECINFO_FLAGS_PR | prot);
>         if (ret < 0) {
>             log_error("%s: failed to accept restrict pages permissions at %#lx: %d", __func__,
> ```





_Suggestion:_
```C
restricted pages
```

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski)

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: all files reviewed, 1 unresolved discussion (waiting on @boryspoplawski)

**Review by mkow — APPROVED** [review]
> Dismissed @boryspoplawski from a discussion.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: :shipit: complete! all files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NM9kwHUE4gBo-WIh-T6:-NMTFHaM8tfSCKQKXGHY:b5kqxwp) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Dismissing Borys from this discussion - it was resolved and Borys was ok with us to dismiss him (he doesn't have much time for Gramine now).

___

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: :shipit: complete! all files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMDc_Xf0J8nn2lWAT_1:-NMTpBHgEKEUUSkw5_J4:b-d5ezs8) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

> Shouldn't the page have still the old permissions until EACCEPTed?

No, the page has the new permissions on this logical processor (the processor which called an OCALL to ask the host OS to perform `EMODPR`). But the local TLBs of other logical processors may still contain the old permissions. And the EACCEPT verifies/forces a TLB flush on this enclave page (some microarchitectural details).
</blockquote></details>

That makes no sense to me. So the untrusted host can add `+x` to memory without enclave consent or even it knowing about that?

___

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1135)* status: :shipit: complete! all files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1135#-NMDc_Xf0J8nn2lWAT_1:-NMTqFA8-u9Tt2FMBmYC:ba5325m) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

That makes no sense to me. So the untrusted host can add `+x` to memory without enclave consent or even it knowing about that?
</blockquote></details>

Ah, this is only restricting memory, nvm
EDMM complexity still amazes me sometimes

___
